src/basic/uid-classification.h

```diff
@@ -33,4 +14,44 @@ typedef struct UGIDAllocationRange {
 int read_login_defs(UGIDAllocationRange *ret_defs, const char *path, const char *root);
 const UGIDAllocationRange *acquire_ugid_allocation_range(void);
 
+typedef enum UGIDRangeFlags {
+        UGID_RANGE_SYSTEM     = 1 << 0,
+        UGID_RANGE_DYNAMIC    = 1 << 1,
+        UGID_RANGE_CONTAINER  = 1 << 2,
+} UGIDRangeFlags;
+
+static inline bool uid_in_range(uid_t uid, UGIDRangeFlags flags) {
+        assert(flags);
+
+        if (FLAGS_SET(flags, UGID_RANGE_DYNAMIC))
+                if (DYNAMIC_UID_MIN <= uid && uid <= DYNAMIC_UID_MAX)
+                        return true;
+
+        if (FLAGS_SET(flags, UGID_RANGE_CONTAINER))
+                if (CONTAINER_UID_BASE_MIN <= uid && uid <= CONTAINER_UID_BASE_MAX)
+                        return true;
+
+        if (FLAGS_SET(flags, UGID_RANGE_SYSTEM)) {
+                const UGIDAllocationRange *defs;
+                assert_se(defs = acquire_ugid_allocation_range());
+
+                if (uid <= defs->system_uid_max)
+                        return true;
+        }
+
+        return false;
+}
+
+static inline bool gid_in_range(gid_t gid, UGIDRangeFlags flags) {
+
+        if (FLAGS_SET(flags, UGID_RANGE_SYSTEM)) {
+                const UGIDAllocationRange *defs;
+                assert_se(defs = acquire_ugid_allocation_range());
+
+                return gid <= defs->system_gid_max;
```

Comment:
this is definitely wrong. It would mean if you pass a dynamic gid, with UGID_RANGE_SYSTEM|UGID_RANGE_DYNAMIC it would say "no".

and in case people configured different limits for system uids and system gids (which is dumb, but the logic behind login.defs is kinda suggesting this), then this would decide something is a system group even though it might have a gid that does not qualify as that because the system gid boundary < system uid boundary.

Reply:
You're right `gid_in_rage()` is incorrect.